### DISCUSSION TITLE
pdk test unit: Assert resource title lengths are compatible with PuppetDB

### DISCUSSION THREAD

**Discussion by Sharpie** [puppetlabs-toy-chest/pct#191] [General]
When PuppetDB stores catalog data in Postgres, it encounters database constraints that may not be present in other parts of the system. A common constraint is that any value inserted into an indexed column must be less than a certain length:

```
Any data type that can be sorted into a well-defined linear order can be indexed by a btree index. The only limitation is that an index entry cannot exceed approximately one-third of a page (after TOAST compression, if applicable).
```

https://www.postgresql.org/docs/11/btree-intro.html

Database pages are nearly always left at the default 8912 bytes long, which means the constraint comes out to 2712 bytes. Long resource titles frequently trigger this constraint as they are covered by many indexes.

A typical error looks like:

```
2021-05-11T23:59:47.225Z ERROR [p.p.command] [14,654,263] [store report] Retrying after attempt 0 for node.hostname.example, due to: org.postgresql.util.PSQLException: ERROR: index row size 2720 exceeds maximum 2712 for index "resource_events_resource_timestamp_20210511z"
```

These errors are often not visible to puppet module developers as they occur long after the code has been written, tested, and then deployed. The PDK can help developers catch this issue early when writing code by providing a unit test that checks resource title lengths.

**Comment by Sharpie** [comment]
This commonly occurs with resources like `notify`, `exec`, or `file` where the long `title` should be moved to a parameter like `message`, `command`, or `path`.

Even though 2712 bytes is a hard limit at which Postgres will refuse to store an entire catalog or report, we may want the test to fail at a much lower value. Each index stores a separate copy of the data and `title` is stored in a good half-dozen to dozen indexes. This means that long titles will bloat the indexes which decreases the efficiency of the database, especially at scale, compared to putting the data in a resource parameter that is not indexed.

A test that flags resource titles longer than 255 characters may be reasonable.

**Comment by DavidS** [comment]
From a content perspective, a limit on resource titles in Puppet DB means a limit on resource titles everywhere, as this would otherwise randomly break content as it moves from one use-case to the next. Shouldn't this then become a basic restriction to be enforced at compile time?

Especially as `pdk test unit` and `pdk validate` are by no means universally used gates into production.

**Comment by Sharpie** [comment]
Enforcing the restriction in Puppet would be a breaking change and so wouldn't take effect until Puppet 8. A test in PDK can be applied to code that runs on older versions. Enforcing a shorter limit, like 255 characters, is also more suited to a test or lint check as these allow for resources to be exempted when needed.
